### PR TITLE
fix: Adjust useChildElementTruncation to eliminate infinite loop issue

### PR DESCRIPTION
**Pull request by eric-briscoe**
### SUMMARY

Fixes edge case where useChildElementTruncation got into infinite loop dues to changes on multiple DOM elements causing a ping pong effect.

This Effort will do refactoring on the useChildElementTruncation file to eliminate this issue.
### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Screen Shot 2022-12-20 at 11 55 21 AM

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts [hunk 1]
```diff
@@ -27,92 +27,65 @@ import { RefObject, useLayoutEffect, useState, useRef } from 'react';
  * (including those completely hidden) and whether any elements
  * are completely hidden.
```
Comment (michael-s-molina) [inline superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts:28]: Can you delete the original description? We don't receive refs anymore.

File: superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts [hunk 2]
```diff
@@ -27,92 +27,65 @@ import { RefObject, useLayoutEffect, useState, useRef } from 'react';
...
-
-        // assures it shows +{number} only when the item is not visible
+      let hiddenElementCounter = 0;
+      childNodes.forEach((node: HTMLElement) => {
+        const itemWidth = node?.offsetWidth;
+        const remainingWidth = clientWidth - width;
+        /**
+         * Evaluate if hiddenElementCounter should increment before adding to width
+         * to ensure we do not count partially hidden / clipped / truncated element as a hidden item
+         */
```
Comment (michael-s-molina) [inline superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts:65]: Just to make it similar to the other comments.
```suggestion
        // Evaluate if hiddenElementCounter should increment before adding to width
        // to ensure we do not count partially hidden / clipped / truncated element as a hidden item
```

File: superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts [hunk 3]
```diff
@@ -27,92 +27,65 @@ import { RefObject, useLayoutEffect, useState, useRef } from 'react';
...
       let width = 0;
-      let hiddenElements = 0;
-      for (let i = 0; i < elementsCount; i += 1) {
-        const itemWidth = (childNodes[i] as HTMLElement).offsetWidth;
-        const remainingWidth = maxWidth - truncationWidth - width - plusSize;
-
-        // assures it shows +{number} only when the item is not visible
+      let hiddenElementCounter = 0;
+      childNodes.forEach((node: HTMLElement) => {
+        const itemWidth = node?.offsetWidth;
```
Comment (michael-s-molina) [inline superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts:60]: If we want to support [transform](https://developer.mozilla.org/en-US/docs/Web/CSS/transform), then [Element.getBoundingClientRect()](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model/Determining_the_dimensions_of_elements#how_much_room_does_it_use_up) would be more suitable.

**Review by kgabryje — CHANGES_REQUESTED** [review]
I'm seeing 2 problems:

1. There's "0" added at the end of `Scope` and `Dependent on` rows in filter cards
2. I see some flaky behaviour in filter cards. I set `test2` to be dependent on `test, test3, test4, test5`. The expected behaviour would be to display `test, test... +2`. You can see at the recording that it sometimes displays `test, test... +1` or `test, test3, ... 0`, neither of which is correct


https://user-images.githubusercontent.com/15073128/209192039-1befdcc6-61a6-462d-be31-9751dcbcd585.mov

**Review by codyml — CHANGES_REQUESTED** [review]
Thanks for this!  If it's safe to do so now, would you mind including a revert of https://github.com/apache/superset/commit/7a94f3afc611993324cb507b23a8f8c320ac76cd in this PR as well to re-enable the deleted column?

**Comment by eric-briscoe** [comment]
> Thanks for this! If it's safe to do so now, would you mind including a revert of [7a94f3a](https://github.com/apache/superset/commit/7a94f3afc611993324cb507b23a8f8c320ac76cd) in this PR as well to re-enable the deleted column?

@codyml I updated to re-enable the disabled column and commented out test case.  Please verify the changes look correct to you.  Thanks!

**Comment by eric-briscoe** [comment]
> I'm seeing 2 problems:
> 
> 1. There's "0" added at the end of `Scope` and `Dependent on` rows in filter cards
> 2. I see some flaky behaviour in filter cards. I set `test2` to be dependent on `test, test3, test4, test5`. The expected behaviour would be to display `test, test... +2`. You can see at the recording that it sometimes displays `test, test... +1` or `test, test3, ... 0`, neither of which is correct
> 
>  Screen.Recording.2022-12-22.at.18.20.32.mov

@kgabryje thank you for catching this.  It took a while to track down the root cause of this.  The behavior has been addressed in recent commit.  The inconsistent behavior had to do with timing on antd popover setting CSS value of display: hidden to the popover component.  While this container has the hidden value all children in DOM will report 0 widths.  Because the CSS value gets set async from our components lifecycle it created odd race conditions and miscalculations.  The current version should show correct and stable results.  Thanks!

File: superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts (hunk 3 above)
Comment (eric-briscoe) [reply superset-frontend/src/hooks/useTruncation/useChildElementTruncation.ts]: @michael-s-molina I changed to use getBoundingClientRect.  Thanks!

**Review by kgabryje — COMMENTED** [review]
I noticed 1 small problem - according to requirements for the truncation logic, `+x` shouldn't include items that are partially visible. For example, if we have `item1, item2, item3` and we display `item1, ite...`, we should display +1, not +2.
Currently it looks like we include that partially visible item in the counter (see screenshot) 
image

**Comment by kgabryje** [comment]
> I noticed 1 small problem - according to requirements for the truncation logic, `+x` shouldn't include items that are partially visible. For example, if we have `item1, item2, item3` and we display `item1, ite...`, we should display +1, not +2. Currently it looks like we include that partially visible item in the counter (see screenshot) image

UPDATE: it seems to be flaky :( When I first open the filter card it says `+2`, when I try a few more times it changes to `+3`

https://user-images.githubusercontent.com/15073128/211848402-b6683f6a-8091-4736-ab3b-041d4f388769.mov

**Comment by rusackas** [comment]
Closing since this isn't likely to be followed up on (cc @eschutho ). Thanks @eric-briscoe ❤️

**Pull request closed without merging**